src/Covenant/Type.hs

```diff
@@ -706,3 +713,132 @@ substitute index toSub = \case
   ThunkT (CompT abstractions xs) ->
     ThunkT . CompT abstractions . fmap (substitute index toSub) $ xs
   BuiltinFlat t -> BuiltinFlat t
+
+-- Pretty printing stuff, somewhat awkward to put it here but doing so avoids orphan instances
+
+-- Keeping the field names for clarify even if we don't use them
+newtype ScopeBoundary = ScopeBoundary {_getBoundary :: Int}
+  deriving newtype (Show, Eq, Ord, Num, Enum)
+
+data PrettyContext (ann :: Type)
+  = PrettyContext {
+      _pcBoundIdents :: Map ScopeBoundary [Doc ann]
+    , _pcCurrentScope :: ScopeBoundary
+    , _pcVarStream :: [Doc ann]
+    }
+
+-- Lazily generated infinite list of variables. Will start with a, b, c... and cycle around to a1, b2, c3 etc.
+-- We could do something more sophisticated but this should work.
+infiniteVars :: forall (ann :: Type). [Doc ann]
+infiniteVars = let aToZ = ['a'..'z']
+                   intStrings = ("" <$ aToZ) <> map (show @Integer) [0..]
+               in zipWith (\x xs -> pretty (x:xs)) aToZ intStrings
+
+-- optics, for convenience. Manually defined b/c TH splices tend to blow up HLS
+boundIdents :: forall (ann :: Type). Lens' (PrettyContext ann) (Map ScopeBoundary [Doc ann])
+boundIdents = lens goGet goSet
+  where
+    goGet (PrettyContext bi _ _) = bi
+    goSet (PrettyContext _ scop vars) bi' = PrettyContext bi' scop vars
+
+currentScope :: forall (ann :: Type). Lens' (PrettyContext ann) ScopeBoundary
+currentScope = lens goGet goSet
+  where
+    goGet (PrettyContext _ scop _) = scop
+    goSet (PrettyContext bi _ vars) scop  = PrettyContext bi scop vars
+
+varStream :: forall (ann :: Type). Lens' (PrettyContext ann) [Doc ann]
+varStream = lens goGet goSet
+  where
+    goGet (PrettyContext _ _ vars) = vars
+    goSet (PrettyContext bi scop _) = PrettyContext bi scop
```

Comment:
And here. It might be worth generating some optics for `PrettyContext` and using those?